### PR TITLE
Fix switch pattern

### PR DESCRIPTION
**Pull request by MarcoRossignoli**
fixes https://github.com/coverlet-coverage/coverlet/issues/936

cc: @daveMueller I did some test and skipping branches and breakpoint expression seems to work

image

image

### CODE REVIEW

**Comment by daveMueller** [comment]
Yes that's a good solution. Thanks again to @jakubch1 for the hint. The only thing that doesn't really fit is the hits count, line 36 is actually only hit once.

image

But it's certainly more lightweight than what I did. I mixed this skip and the other algorithm and therefore had to do some changes in the instrumenter ..., definitely more heavy.

**Comment by MarcoRossignoli** [comment]
@daveMueller with the idea of "skip nested lines" works better
image

**Comment by MarcoRossignoli** [comment]
@daveMueller please feel free to review this.

**Comment by daveMueller** [comment]
@MarcoRossignoli Sure I will just need approx. 2 days to find some spare time. I'm really interested in the key differences to my implementation and why I couldn't make I run without touching the ```Intrumenter```. Also I think this might solve another issue I have in mind.

File: src/coverlet.core/Symbols/CecilSymbolHelper.cs [hunk 1]
```diff
@@ -714,6 +729,43 @@ static Instruction GetPreviousNoNopInstruction(Instruction i)
...
+               IL_0001: brtrue.s IL_0004
+               // if (value is int)
+               // sequence point: (line 34, col 9) to (line 40, col 10) in C:\git\coverletfork\test\coverlet.core.tests\Samples\Instrumentation.SelectionStatements.cs
+               IL_0003: nop
+                // sequence point: hidden
+                ...
+             */
+            if (
+                    instruction.OpCode == OpCodes.Ldc_I4 && instruction.Operand is int operandValue && operandValue == 1 &&
+                    instruction.Next?.OpCode == OpCodes.Brtrue &&
```
Comment (jakubch1) [inline src/coverlet.core/Symbols/CecilSymbolHelper.cs:752]: To check exactly above pattern you should also check that target of `brtrue.s` is instruction after `nop`

Comment (MarcoRossignoli) [reply src/coverlet.core/Symbols/CecilSymbolHelper.cs:752]: Uh you're right thanks.

Comment (daveMueller) [reply src/coverlet.core/Symbols/CecilSymbolHelper.cs:752]: According to the pattern I also checked ```methodDefinition.DebugInformation.GetSequencePoint(instruction.Next?.Next?.Next).IsHidden```. That's also what I observed in the IL that after the ```nop``` instruction a hidden sequence point follows. But I'm not a specialist here and maybe this isn't important.

File: src/coverlet.core/Coverage.cs [hunk 2]
```diff
@@ -392,7 +392,33 @@ private void CalculateCoverage()
...
+                    {
+                        continue;
+                    }
+
+                    foreach (HitCandidate hitCandidateToCompare in result.HitCandidates)
+                    {
+                        if (hitCandidate != hitCandidateToCompare && !hitCandidateToCompare.isBranch)
+                        {
+                            if (hitCandidateToCompare.start >= hitCandidate.start &&
+                               hitCandidateToCompare.end <= hitCandidate.end)
```
Comment (daveMueller) [inline src/coverlet.core/Coverage.cs:409]: Here is where I made the mistake in my branch. I just had an issue with the range :man_facepalming:. I really like your idea of saving the excluded lines in a ```HashSet<int>``` this removes the overhead of finding the top most :thumbsup:.

**Pull request merged**